// @magicyan/discord/dist/functions/embeds/embedplus.cjs
'use strict';

const discord_js = require('discord.js');
const chars = require('../../constants/chars.cjs');
const assets = require('./assets.cjs');
const fields = require('./fields.cjs');
const footer = require('./footer.cjs');

var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
class EmbedPlusBuilder extends discord_js.EmbedBuilder {
  constructor(data) {
    const { mergeFields = false, extends: extendsEmbed, ...embedData } = data;
    const extendsEmbedData = extendsEmbed ? "data" in extendsEmbed ? extendsEmbed.data : extendsEmbed : {};
    const { fields: extendsFields, ...extendsData } = extendsEmbedData;
    const fields$1 = (mergeFields ? [extendsFields ?? [], data.fields ?? []].flat() : data.fields ?? extendsFields ?? []).map((field) => Object.assign(
      { name: field.name ?? chars.chars.invisible, value: field.value ?? chars.chars.invisible },
      field.inline !== void 0 ? { inline: field.inline } : {}
    ));
    const builderData = Object.assign({}, extendsData, embedData, { fields: fields$1 });
    const { color, footer: footer$1, image, thumbnail, timestamp } = embedData;
    if (footer$1)
      Object.assign(builderData, { footer: footer.createEmbedFooter(footer$1) });
    if (image)
      Object.assign(builderData, { image: assets.createEmbedAsset(image) });
    if (thumbnail)
      Object.assign(builderData, { thumbnail: assets.createEmbedAsset(thumbnail) });
    const embed = new discord_js.EmbedBuilder(builderData);
    if (timestamp)
      embed.setTimestamp(
        typeof timestamp === "string" ? new Date(timestamp) : timestamp
      );
    if (color)
      embed.setColor(color);
    super(embed.data);
    __publicField(this, "fields");
    this.fields = new fields.EmbedPlusFields(this);
  }
  update(data) {
    const updated = createEmbed({ mergeFields: true, extends: this, ...data });
    Object.assign(this.data, updated.data);
    return this;
  }
  has(property) {
    return Boolean(this.data[property]);
  }
  toArray() {
    return Array.from([this]);
  }
  toString(space = 2) {
    return JSON.stringify(this, null, space);
  }
  setBorderColor(color) {
    if (color === null) {
      this.setColor("#2B2D31");
    } else if (typeof color === "number") {
      this.update({ color });
    } else {
      this.setColor(color);
    }
    return this;
  }
  setAsset(asset, source) {
    if (source === null) {
      asset === "image" ? this.setImage(source) : this.setThumbnail(source);
    } else {
      this.update({ [asset]: source });
    }
    return this;
  }
  static fromInteraction(interaction, index = 0, data = {}) {
    return EmbedPlusBuilder.fromMessage(interaction.message, index, data);
  }
  static fromMessage(message, index = 0, data = {}) {
    return new EmbedPlusBuilder(Object.assign({ extends: message.embeds[index] }, data));
  }
}
function createEmbed(options) {
  const { array = false, interaction, ...data } = options;
  const embed = interaction ? EmbedPlusBuilder.fromInteraction(interaction, 0, data) : new EmbedPlusBuilder(data);
  return array ? [embed] : embed;
}

exports.EmbedPlusBuilder = EmbedPlusBuilder;
exports.createEmbed = createEmbed;
